hoglet.go

```diff
@@ -3,114 +3,170 @@ package hoglet
 import (
 	"context"
 	"errors"
-	"sync"
+	"sync/atomic"
+	"time"
 )
 
-// Breaker wraps a function and behaves like a circuit breaker: it triggers (opens) when the wrapped function fails and
-// stops calling the wrapped function until it closes again, returning [ErrBreakerOpen] in the meantime.
+// Circuit wraps a function and behaves like a simple circuit and breaker: it opens when the wrapped function fails and
+// stops calling the wrapped function until it closes again, returning [ErrCircuitOpen] in the meantime.
 //
-// A zero breaker will not panic, but wraps a noop. Use [NewBreaker] instead.
-type Breaker[IN, OUT any] struct {
+// A zero circuit will not panic, but wraps a noop. Use [NewCircuit] instead.
+type Circuit[IN, OUT any] struct {
 	f       BreakableFunc[IN, OUT]
-	trigger Trigger
-	options options
+	breaker Breaker
+	options
+
+	// State
+
+	openedAt atomic.Int64 // unix microseconds
 }
 
-// We use a sub-struct to avoid requiring type parameters in [Option].
+// options is a sub-struct to avoid requiring type parameters in the [Option] type.
 type options struct {
 	// isFailure is a filter function that determines whether an error can open the breaker.
 	isFailure func(error) bool
+
+	// halfOpenDelay is the duration the circuit will stay open before switching to the half-open state, where a
+	// limited (~1) amount of calls are allowed that - if successful - may re-close the breaker.
+	halfOpenDelay time.Duration
 }
 
-// Trigger is the interface implemented by the different
-type Trigger interface {
-	// State returns the current state of the breaker.
-	// It is called exactly once per call to [Do], before calling the wrapped function.
-	State() State
-	// Observe will only be called if [State] returns StateOpen or StateHalfOpen.
-	// It is called at most once per call to [Do], after calling the wrapped function, depending on the breaker state.
-	Observe(failure bool)
+// Breaker is the interface implemented by the different breakers, responsible for actually opening the circuit.
+// Each implementation behaves differently when deciding whether to open the breaker upon failure.
+type Breaker interface {
```

Comment:
Do we need to export this if the functions are private?

Reply:
we don't really have to, but since it's a parameter to NewCircuit, I think it's a bit more ergonomic to have it exported.

Most IDEs can deal with it not being exported, but I think it's still a bit safer?

Reply:
Ok let's leave it like that